src/ucp/proto/proto_perf.c

```diff
@@ -0,0 +1,400 @@
+};
+
+/*
+ * Protocol performance structure.
+ */
+struct ucp_proto_perf {
+    /* Name of the protocol */
+    char            name[UCP_PROTO_DESC_STR_MAX];
+
+    /* List of segments */
+    ucs_list_link_t segments;
+};
+
+/*
+ * Iterate over the segments of the protocol perf structure function.
+ */
+#define ucp_proto_perf_segment_foreach(_perf, _seg) \
+    ucs_list_for_each((_seg), &(_perf)->segments, list)
+
+static const char *ucp_proto_perf_factor_names[] = {
+    [UCP_PROTO_PERF_FACTOR_LOCAL_CPU]  = "cpu",
+    [UCP_PROTO_PERF_FACTOR_REMOTE_CPU] = "rem_cpu",
+    [UCP_PROTO_PERF_FACTOR_TL_SEND]    = "send",
+    [UCP_PROTO_PERF_FACTOR_TL_RECEIVE] = "recv",
+    [UCP_PROTO_PERF_FACTOR_LATENCY]    = "lat",
+    [UCP_PROTO_PERF_FACTOR_SINGLE]     = "sngl",
+    [UCP_PROTO_PERF_FACTOR_MULTI]      = "mult",
+    [UCP_PROTO_PERF_FACTOR_LAST]       = NULL
+};
+
+static void ucp_proto_perf_check(ucp_proto_perf_t *perf)
+{
+#if ENABLE_ASSERT
+    ucp_proto_perf_segment_t *seg;
+    size_t min_start;
+
+    min_start = 0;
+    ucp_proto_perf_segment_foreach(perf, seg) {
+        ucs_assertv((seg->start >= min_start) && (seg->start <= seg->end),
+                    "perf=%p seg->start=%zu seg->end=%zu min_start=%zu", perf,
+                    seg->start, seg->end, min_start);
+        min_start = seg->end + 1;
+    }
+#endif
+}
+
+static ucs_status_t ucp_proto_perf_segment_new(const ucp_proto_perf_t *perf,
+                                               size_t start, size_t end,
+                                               ucp_proto_perf_segment_t **seg_p)
+{
+    ucp_proto_perf_factor_id_t factor_id;
+    ucp_proto_perf_segment_t *seg;
+
+    ucs_assertv(start <= end, "perf=%p start=%zu end=%zu", perf, start, end);
+
+    seg = ucs_malloc(sizeof(*seg), "ucp_proto_perf_segment");
+    if (seg == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    seg->start = start;
+    seg->end   = end;
+    seg->node  = NULL;
+    for (factor_id = 0; factor_id < UCP_PROTO_PERF_FACTOR_LAST; factor_id++) {
+        seg->factors[factor_id] = UCS_LINEAR_FUNC_ZERO;
+    }
+
+    *seg_p = seg;
+    return UCS_OK;
+}
+
+static ucs_status_t ucp_proto_perf_segment_split(const ucp_proto_perf_t *perf,
+                                                 ucp_proto_perf_segment_t *seg,
+                                                 size_t seg_end)
+{
+    ucp_proto_perf_factor_id_t factor_id;
+    ucp_proto_perf_segment_t *new_seg;
+    ucs_status_t status;
+
+    ucs_assertv(seg_end < seg->end, "seg=%p seg->end=%zu seg_end=%zu", seg,
+                seg->end, seg_end);
+
+    status = ucp_proto_perf_segment_new(perf, seg_end + 1, seg->end, &new_seg);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    for (factor_id = 0; factor_id < UCP_PROTO_PERF_FACTOR_LAST; factor_id++) {
+        new_seg->factors[factor_id] = seg->factors[factor_id];
+    }
+    new_seg->node = ucp_proto_perf_node_dup(seg->node);
```

Comment:
we might need to handle error path like using `goto fail` or there is risk to silently endup with incomplete perf segments? there are other instances.

Reply:
w.r.t perf_node, a failure to create it is not considered fatal (because perf tree is for debug); in case of failure it will remain NULL and ignored later on.

Reply:
I see, I assume there is no ambiguity ontroduced in the dump for perf segments/nodes.